link-grammar/parse/fast-match.c

```diff
@@ -197,47 +127,103 @@ static Match_node **get_match_table_entry(unsigned int size, Match_node **t,
 		}
 	}
 
 	return &t[h];
 }
 
 /**
- * The disjunct d (whose left or right pointer points to c) is put
- * into the appropriate hash table
- * dir =  1, we're putting this into a right table.
- * dir = -1, we're putting this into a left table.
+ * Add the match nodes at sbin to the table entry of their shallow connector.
  */
-static void put_into_match_table(Sentence sent, unsigned int size,
-                                 Match_node ** t, Disjunct * d,
-                                 Connector * c, int dir)
+static void add_to_table_entry(unsigned int tsize, Match_node **table,
+                               int dir, sortbin *sbin)
 {
-	Match_node *m, **xl;
+	Match_node *m_next;
+	for (Match_node *m = sbin->head; NULL != m; m = m_next)
+	{
+		Connector *c = (0 == dir) ? m->d->left : m->d->right;
+		assert(NULL != c);
 
-	m = pool_alloc(sent->fm_Match_node);
-	m->next = NULL;
-	m->d = d;
+		Match_node **xl = get_match_table_entry(tsize, table, c, dir);
+		assert(&match_list_not_found != xl, "get_match_table_entry: Overflow");
 
-	xl = get_match_table_entry(size, t, c, dir);
-	assert(NULL != xl, "get_match_table_entry: Overflow");
-	if (dir == 1) {
-		*xl = add_to_right_table_list(m, *xl);
+		m_next = m->next; /* Remember before overwriting. */
+
+		/* Insert m at head of list */
+		m->next = *xl;
+		*xl = m;
+	}
+}
+
+/**
+ * Put the nearest_word-sorted disjuncts at sbin into the given hash table.
+ * Since add_to_table_entry() inserts at entry's head, insert in the
+ * reverse order than the needed match-list order of nearest_word.
+ *
+ * @param tsize The hash table size.
+ * @param table The hash table.
+ * @param w The word to which these disjuncts belong.
+ * @param dir 0: Put it into a right table; 1: Put it  into a left table.
+ * @param sbin Indexed by the disjuncts' shallow-connector nearest_word.
+ * @param sent_length Sentence length, for sbin index upper bound.
+ */
+static void put_into_match_table(unsigned int tsize, Match_node **table,
+                                 int w, int dir, sortbin *sbin,
+                                 size_t sent_length)
+{
+	if (0 == dir)
+	{
+		/* Left match-list needed order: Increasing nearest_word. */
+		for (WordIdx sbw = 0; sbw < (WordIdx)w; sbw++)
+		{
+			add_to_table_entry(tsize, table, dir, &sbin[sbw]);
+		}
 	}
 	else
 	{
-		*xl = add_to_left_table_list(m, *xl);
+		/* Right match-list needed order: Decreasing nearest_word. */
+		for (WordIdx sbw = sent_length-1; sbw > (WordIdx)w; sbw--)
+		{
+			add_to_table_entry(tsize, table, dir, &sbin[sbw]);
+		}
 	}
 }
 
-fast_matcher_t* alloc_fast_matcher(const Sentence sent)
+static void clean_sortbin(sortbin *sbin, size_t sent_length)
+{
+	for (unsigned int i = 0; i < sent_length; i++)
+				sbin[i].head = NULL;
+}
+
+/**
+ * Put the disjunct d (whose left or right pointer points to c) into
+ * a list according to the c->nearest_word. This is a bucket sort.
+ */
+static void sort_by_nearest_word(Match_node *m, sortbin *sbin)
```

Comment:
this is misnamed; it's not sorting anything, its just prepending to a linked-list.

Reply:
The sorting is implemented by just assigning to the bucket.
However, the bucket selection is part of the sort, and it is done by the caller.
So I propose this change to make it clearer and the name appropriate:
``` diff
-                      sort_by_nearest_word(m, &sbin[d->left->nearest_word]);
+                      sort_by_nearest_word(m, sbin, d->left->nearest_word);
...
-                      sort_by_nearest_word(m, &sbin[d->right->nearest_word]);
+                      sort_by_nearest_word(m, sbin, d->right->nearest_word);
...
 /**
- * Put the disjunct d (whose left or right pointer points to c) into
- * a list according to the c->nearest_word. This is a bucket sort.
+ * Put the Match_node m into a list according to nearest_word.
+ * This is a bucket sort.
  */
-static void sort_by_nearest_word(Match_node *m, sortbin *sbin)
+static void sort_by_nearest_word(Match_node *m, sortbin *sbin, int nearest_word)
 {
 
+       sbin = &sbin[nearest_word]; /* Bucket selection. */
+
 #if ML_COMPAT
```

Above, I added the bucket selection to the function, making it a real sorting.
I also fixed the function comment (which I forgot to change) and fixed the indentation.
I will send a PR for that.